fish-rust/src/builtins/ulimit.rs

```diff
@@ -0,0 +1,354 @@
+use std::cmp::max;
+
+use libc::{c_int, RLIM_INFINITY};
+use nix::{
+    errno::Errno::EPERM,
+    sys::resource::{self, getrlimit, setrlimit},
+};
+
+use crate::{
+    builtins::shared::STATUS_CMD_ERROR,
+    ffi::{fish_wcswidth2, parser_t},
+    wchar::{wstr, L},
+    wchar_ffi::WCharToFFI,
+    wgetopt::{wgetopter_t, wopt, woption, woption_argument_t::no_argument},
+    wutil::wgettext_fmt,
+};
+
+use super::shared::{builtin_wperror, io_streams_t, STATUS_CMD_OK};
+
+/// Struct describing a resource limit.
+#[derive(Debug, Copy, Clone)]
+struct Resource<'a> {
+    /// resource ID
+    resource: resource::Resource,
+    /// description of resource
+    desc: &'a wstr,
+    /// switch used on commandline to specify resource
+    switch_char: char,
+    /// the implicit multiplier used when setting getting values
+    multiplier: i32,
+}
+
+/// Array of `Resource` structs, describing all known resource types.
+const resource_arr: &[Resource] = &[
+    #[cfg(any(target_os = "freebsd", target_os = "dragonfly"))]
+    Resource {
+        resource: resource::Resource::RLIMIT_SBSIZE,
+        desc: L!("Maximum size of socket buffers"),
+        switch_char: 'b',
+        multiplier: 1024,
+    },
+    Resource {
+        resource: resource::Resource::RLIMIT_CORE,
+        desc: L!("Maximum size of core files created"),
+        switch_char: 'c',
+        multiplier: 1024,
+    },
+    Resource {
+        resource: resource::Resource::RLIMIT_DATA,
+        desc: L!("Maximum size of a process’s data segment"),
+        switch_char: 'd',
+        multiplier: 1024,
+    },
+    #[cfg(any(target_os = "android", target_os = "linux"))]
+    Resource {
+        resource: resource::Resource::RLIMIT_NICE,
+        desc: L!("Control of maximum nice priority"),
+        switch_char: 'e',
+        multiplier: 1,
+    },
+    Resource {
+        resource: resource::Resource::RLIMIT_FSIZE,
+        desc: L!("Maximum size of files created by the shell"),
+        switch_char: 'f',
+        multiplier: 1024,
+    },
+    #[cfg(any(target_os = "android", target_os = "linux"))]
+    Resource {
+        resource: resource::Resource::RLIMIT_SIGPENDING,
+        desc: L!("Maximum number of pending signals"),
+        switch_char: 'i',
+        multiplier: 1,
+    },
+    #[cfg(any(
+        target_os = "android",
+        target_os = "freebsd",
+        target_os = "openbsd",
+        target_os = "linux",
+        target_os = "netbsd"
+    ))]
+    Resource {
+        resource: resource::Resource::RLIMIT_MEMLOCK,
+        desc: L!("Maximum size that may be locked into memory"),
+        switch_char: 'l',
+        multiplier: 1024,
+    },
+    #[cfg(any(
+        target_os = "android",
+        target_os = "freebsd",
+        target_os = "netbsd",
+        target_os = "openbsd",
+        target_os = "linux",
+    ))]
```

Comment:
You're missing dragonflybsd here, though you have it elsewhere. I'm introducing a build-time feature `bsd` that groups all BSDs together, and can be used instead of micro-targeting the exact os name for BSD derivatives. 

I'll merge that into master soon so you can rebase on top of it and take advantage of it for anything that isn't actually FreeBSD-exclusive.

This is, of course, still not doing feature detection but it'll make it easier not to miss things.

Is the lack of macOS here intentional due to a missing symbol in the libc or nix crates? macOS supports RLIMIT_RSS (I think any \*nix would).